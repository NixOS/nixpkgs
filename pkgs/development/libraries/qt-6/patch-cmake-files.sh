#! /bin/sh

# patch files

# first argument is "patch"
# first group of arguments are key=val pairs
# group delimiter: --
# second group of arguments are files to patch

# perl must be in $PATH

set -eu

log() {
  echo "$0: $*" >&2
}

if [ $# = 0 ] || [ "$1" != patch ]; then
  log "fatal error: first argument must be 'patch'"
  log "hint: this script should not be in buildInputs"
  exit 1
fi
shift

log "arguments: $*" # debug

for arg in "$@"
do
  case "$arg" in
    *=*) log "eval $arg"; eval "$arg"; shift;;
    --) shift; break;;
    *) log "fatal error: bad argument '$arg'. expected key=val or --"; exit 1;;
  esac
done

if [ $# = 0 ]; then
  log "no files to patch -> ignore"
  exit 0
fi

log "patching $# files"

cat <<EOF
$0: debug:
  REGEX_FILE = $REGEX_FILE
  QT_COMPAT_VERSION = $QT_COMPAT_VERSION
  QT_MODULE_NAME = $QT_MODULE_NAME
  NIX_OUT_PATH = $NIX_OUT_PATH
  NIX_DEV_PATH = $NIX_DEV_PATH
  NIX_BIN_PATH = $NIX_BIN_PATH
EOF

if ! [ -e "$REGEX_FILE" ]; then
  log "fatal error: missing REGEX_FILE: $REGEX_FILE"
  exit 1
fi

log "build regex"

escape_b() {
  echo "$1" | sed 's,/,\\/,g'
}

regex="$( ( echo; cat "$REGEX_FILE"; echo ) | perl -p -0 -e '
  s/\n#[^\n]*/\n/g;
  s/\(\(QT_COMPAT_VERSION\)\)/'"$QT_COMPAT_VERSION"'/g;
  s/\(\(QT_MODULE_NAME\)\)/'"$QT_MODULE_NAME"'/g;
  s/\(\(NIX_OUT_PATH\)\)/'"$(escape_b "$NIX_OUT_PATH")"'/g;
  s/\(\(NIX_DEV_PATH\)\)/'"$(escape_b "$NIX_DEV_PATH")"'/g;
  s/\(\(NIX_BIN_PATH\)\)/'"$(escape_b "$NIX_BIN_PATH")"'/g;
  s/([^\n])\//\1\\\//g;
  s/\n-([^\n]+)\n\+([^\n]*)(?:\n\/([^\n]+))?\n/s\/\1\/\2\/\3;\n/g
')"

log "regex:"; echo "$regex" # debug

#log "$# input files:"; for f in "$@"; do echo "  $f"; done # debug

exec perl -00 -p -i -e "$regex" "$@"
