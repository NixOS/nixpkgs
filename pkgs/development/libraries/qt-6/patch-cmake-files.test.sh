#! /bin/sh

temp_file=$(mktemp)

cp -v "patch-cmake-files.test.input.txt" "$temp_file"

echo "$0: patching $temp_file"

sh patch-cmake-files.sh patch \
  REGEX_FILE=./patch-cmake-files.regex.diff \
  QT_COMPAT_VERSION=6.2.2 \
  QT_MODULE_NAME=QTBASE \
  NIX_OUT_PATH=/nix/store/xxxx-out \
  NIX_DEV_PATH=/nix/store/xxxx-dev \
  NIX_BIN_PATH=/nix/store/xxxx-bin \
  -- \
  "$temp_file"

echo "$0: done patching $temp_file"

expected_file="patch-cmake-files.test.expected.txt"

echo
echo "comparing ..."
echo "a: $expected_file"
echo "b: $temp_file"
if diff -u --color=always "$expected_file" "$temp_file"
then echo ok
else
  echo "mismatch. to update the expected file:"
  echo "cp $temp_file $expected_file"
fi
