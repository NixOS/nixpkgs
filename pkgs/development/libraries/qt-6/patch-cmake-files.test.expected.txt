# set all paths in file header
set(_QTBASE_NIX_OUT "/nix/store/xxxx-out")
set(_QTBASE_NIX_DEV "/nix/store/xxxx-dev")
set(_QTBASE_NIX_BIN "/nix/store/xxxx-bin")

# unset all paths in file footer
set(_QTBASE_NIX_OUT)
set(_QTBASE_NIX_DEV)
set(_QTBASE_NIX_BIN)

# include
${_QTBASE_NIX_DEV}/include
${_QTBASE_NIX_DEV}/include

# libexec
${_QTBASE_NIX_OUT}/libexec
${_QTBASE_NIX_OUT}/libexec
/nix/store/xxxx-out/libexec

# qtbase libexec
${QtBase_BINARY_DIR}/libexec

# lib
${_QTBASE_NIX_OUT}/lib
${_QTBASE_NIX_OUT}/lib
/nix/store/xxxx-out/lib

# plugins
${_QTBASE_NIX_BIN}/lib/qt-6.2.2/plugins
${_QTBASE_NIX_BIN}/lib/qt-6.2.2/plugins

# bin
# all qt binaries are devTools
${_QTBASE_NIX_DEV}/bin
${_QTBASE_NIX_DEV}/bin
/nix/store/xxxx-dev/bin

# mkspecs
${_QTBASE_NIX_DEV}/mkspecs
${_QTBASE_NIX_DEV}/mkspecs
/nix/store/xxxx-dev/mkspecs
/nix/store/xxxx-dev/mkspecs
/nix/store/xxxx-dev/mkspecs
/nix/store/xxxx-dev/mkspecs
/nix/store/xxxx-dev/mkspecs

# qml
${_QTBASE_NIX_OUT}/lib/qt-6.2.2/lib/qml
${_QTBASE_NIX_OUT}/lib/qt-6.2.2/lib/qml

# doc
/nix/store/xxxx-out/share/doc

# add INTERFACE_LINK_DIRECTORIES
(  INTERFACE_LINK_DIRECTORIES "${_QTBASE_NIX_OUT}/lib"
  INTERFACE_INCLUDE_DIRECTORIES )
